lighthouse-core/test/audits/seo/hreflang-test.js

```diff
@@ -76,15 +76,58 @@ describe('SEO: Document has valid hreflang code', () => {
   it('returns all failing items', () => {
     const artifacts = {
       LinkElements: [
-        {rel: 'alternate', hreflang: 'xx1', href: 'http://xx1.example.com/', source: 'headers'},
-        {rel: 'alternate', hreflang: 'xx2', href: 'http://xx2.example.com/', source: 'headers'},
-        {rel: 'alternate', hreflang: 'xx3', href: 'http://xx3.example.com/', source: 'head'},
-        {rel: 'alternate', hreflang: 'xx4', href: 'http://xx4.example.com/', source: 'head'},
+        {rel: 'alternate', hreflang: 'xx1', hrefRaw: 'http://xx1.example.com/', source: 'headers'},
+        {rel: 'alternate', hreflang: 'xx2', hrefRaw: 'http://xx2.example.com/', source: 'headers'},
+        {rel: 'alternate', hreflang: 'xx3', hrefRaw: 'http://xx3.example.com/', source: 'head'},
+        {rel: 'alternate', hreflang: 'xx4', hrefRaw: 'http://xx4.example.com/', source: 'head'},
+      ],
+    };
+
+    const {score, details} = HreflangAudit.audit(artifacts);
+    assert.equal(score, 0);
+    assert.equal(details.items.length, 4);
+  });
+
+  it('fails when the hreflang url is not fully-qualified', () => {
+    const artifacts = {
+      LinkElements: [
+        {rel: 'alternate', hreflang: 'es', hrefRaw: 'example.com', source: 'head'},
+        {rel: 'alternate', hreflang: 'es', hrefRaw: '//example.com', source: 'headers'},
       ],
     };
 
-    const auditResult = HreflangAudit.audit(artifacts);
-    assert.equal(auditResult.score, 0);
-    assert.equal(auditResult.details.items.length, 4);
+    const {score, details} = HreflangAudit.audit(artifacts);
+    assert.equal(score, 0);
+    assert.equal(details.items.length, 2);
+  });
+
+  it('fails with an invalid language code and a href which is not fully-qualified', () => {
+    const artifacts = {
+      LinkElements: [
+        {rel: 'alternate', hreflang: ' es', hrefRaw: 'example.com', source: 'head'},
+        {rel: 'alternate', hreflang: 'xx1', hrefRaw: '//example.com', source: 'headers'},
+      ],
+    };
+
+    const {score} = HreflangAudit.audit(artifacts);
+    assert.equal(score, 0);
+  });
+
+  it('outputs the reasons for which a hreflang failed', () => {
+    const artifacts = {
+      LinkElements: [
+        {rel: 'alternate', hreflang: '@@', hrefRaw: 'example.com', source: 'head'},
+        {rel: 'alternate', hreflang: 'fr', hrefRaw: 'example.com', source: 'head'},
+        {rel: 'alternate', hreflang: '@@', hrefRaw: 'https://example.com', source: 'head'},
+        {rel: 'alternate', hreflang: 'fr', hrefRaw: 'https://example.com', source: 'head'},
+      ],
+    };
+
+    const {details: {items}} = HreflangAudit.audit(artifacts);
+
+    assert.equal(items.length, 3);
+    assert.equal(items[0].subItems.items.length, 2);
```

Comment:
we should assert `expect(items[0].subItems.items[0].reason).toBeDisplayString('The message')` to catch the `str_` issue 👍

Reply:
Is the thing to test for where I didn't wrap this in `str_()`?: `reasons.push(UIStrings.unexpectedLanguage)`

Might have misunderstood but `expect(reason).toBeDisplayString('Unexpected language code')` will pass regardless of wrapping the message in `str_()` or not, is it an idea to mock `i18n.createMessageInstanceIdFn()` and assert it was invoked the right number of times or something?

Reply:
> expect(reason).toBeDisplayString('Unexpected language code') will pass regardless of wrapping the message in str_() or not

Ah, dang you're right. Nevermind then!

We'll have to figure out as a team the best way to try to automatically catch these types of errors.

Reply:
just as an FYI for future contributions @umaar , `expect(thing).toHaveLength(n)` provides a much more debuggable error message than the basic assert on the length since it will show what the array actually was, not different in the case of empty array, but if it's off by 1 you get to see what was missing, etc :)